include/arch/arm/arch.h

```diff
@@ -38,6 +38,7 @@
 #elif defined(CONFIG_CPU_CORTEX_R)
 #include <arch/arm/cortex_r/cpu.h>
 #include <arch/arm/cortex_r/sys_io.h>
+#include <arch/arm/cortex_r/timer.h>
```

Comment:
The Generic Timer is defined as an optional extension to both ARMv7-A and -R, as well as ARMv8-A and -R. The same set of CP15 registers apply to both -A and -R. Is there any specific reason you made `timer.h` as a Cortex-R specific header?

IMHO, the Generic Timer CP15 access functions should really be defined inside the system timer driver. By the way, these register access functions are actually defined in CMSIS-Core(A) and CMSIS-Core(R) (the latter is experimental), so using those functions could also be an option.

Reply:
> As for ARMv8-A, we could either patch CMSIS-Core(A), or write a separate driver header that maps to CMSIS Generic Timer-related functions.

Given that CMSIS-Core(A) is only for Cortex-A5/A7/A9 why do we want to patch it to support Cortex-A53 & co? That could be more time wasting that just accessing directly the registers in the helper functions. With Armv8 I'm just using:
```
__asm__ __volatile__("msr...
__asm__ __volatile__("mrs %0, ...
```

>  Is there any specific reason you made timer.h as a Cortex-R specific header?

As I specified the commit regarding the Cortex-R was there just to explain how this driver needs to be used. We can drop it.

> IMHO, the Generic Timer CP15 access functions should really be defined inside the system timer driver.

This is not going to work with AArch64 where we directly use `msr/mrs` to access the registers.

Reply:
> This is not going to work with AArch64 where we directly use `msr/mrs` to access the registers.

Didn't realise AArch64 was doing something completely different; I can't really insist on using CMSIS then.